crates/ika-core/src/dwallet_mpc/crytographic_computation/mpc_computations/presign.rs

```
@@ -4,66 +4,306 @@
+
+        let public_input: <PresignParty<Secp256K1ECDSAProtocol> as mpc::Party>::PublicInput =
+            (protocol_public_parameters, decentralized_party_dkg_output).into();
+        Ok(PresignPublicInputByProtocol::Secp256k1ECDSA(public_input))
+    }
+
+    pub(crate) fn try_new_v2(
+        protocol: DWalletSignatureScheme,
+        versioned_network_encryption_key_public_data: &VersionedNetworkEncryptionKeyPublicData,
+        dwallet_public_output: Option<MPCPublicOutput>,
+    ) -> DwalletMPCResult<Self> {
+        let input = match protocol {
+            DWalletSignatureScheme::ECDSASecp256k1 => {
+                let protocol_public_parameters = versioned_network_encryption_key_public_data
+                    .secp256k1_protocol_public_parameters();
+
+                let deserialized_dwallet_public_output =
+                    dwallet_public_output.and_then(|bytes| bcs::from_bytes(&bytes).ok());
+                let pub_input =
+                    <PresignParty<Secp256K1ECDSAProtocol> as mpc::Party>::PublicInput::from((
+                        protocol_public_parameters,
+                        deserialized_dwallet_public_output,
+                    ));
+                PresignPublicInputByProtocol::Secp256k1ECDSA(pub_input)
+            }
+            DWalletSignatureScheme::SchnorrkelSubstrate => {
+                let protocol_public_parameters = versioned_network_encryption_key_public_data
+                    .ristretto_protocol_public_parameters()?;
+                let deserialized_dwallet_public_output: Option<
+                    <RistrettoAsyncDKGProtocol as dkg::Protocol>::DecentralizedPartyDKGOutput,
+                > = dwallet_public_output.and_then(|bytes| bcs::from_bytes(&bytes).ok());
+                let pub_input =
+                    <PresignParty<RistrettoSchnorrkelSubstrateProtocol> as mpc::Party>::PublicInput::from((protocol_public_parameters, deserialized_dwallet_public_output));
```

deserialization is good, but how is the code compiling, if the `from()` expects `Option<Self::DecentralizedPartyTargetedDKGOutput>`?

---

Also the `and_then` and `.ok()` is bad here, we should return an error and reject session in case the deserialization failed